sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -14,197 +16,180 @@ from ...cluster._hierarchical_fast cimport UnionFind
 from ...utils._typedefs cimport ITYPE_t, DTYPE_t
 from ...utils._typedefs import ITYPE, DTYPE
 
-cpdef cnp.ndarray[cnp.double_t, ndim=2] mst_from_distance_matrix(
-    cnp.ndarray[cnp.double_t, ndim=2] distance_matrix
+# Numpy structured dtype representing a single ordered edge in Prim's algorithm
+MST_edge_dtype = np.dtype([
+    ("current_node", np.intp),
+    ("next_node", np.intp),
+    ("distance", np.float64),
+])
+
+# Packed shouldn't make a difference since they're all 8-byte quantities,
+# but it's included just to be safe.
+ctypedef packed struct MST_edge_t:
+    cnp.intp_t current_node
+    cnp.intp_t next_node
+    cnp.float64_t distance
```

Comment:
This is much more convenient, thanks!

Yet, according to [the specification of `npy_intp`](https://numpy.org/doc/stable/reference/c-api/dtype.html#c.npy_intp) (aliased by `np.intp_t), `current_node` will be stored on 4 bytes on 32bit architecture. I think it should be safe though.

Looking at definition of custom types and `dtype` in `_typedefs.pyx`:
https://github.com/scikit-learn/scikit-learn/blob/b157ac718754a235e443f3f5671fda7d2cd1dbba/sklearn/utils/_typedefs.pyx#L7-L13

defining `MST_edge_dtype` at runtime might be doable using something like:

```python
cdef MST_edge_t MST_edge_placeholder
cdef MST_edge_t[:] MST_edge_placeholder_view = (
    <MST_edge_placeholder[:1]> &MST_edge_placeholder
)
MST_edge_dtype = np.asarray(MST_edge_placeholder_view).dtype
```

But I am unsure this works entirely for composite types: probably fields can't accessed in Python via their names (e.g. using `min_spanning_tree["distance"]`)?

Reply:
Since we never rely on signed indexing (and indeed a negative index here would be problematic), I went ahead and updated to use `int64` explicitly. That should alleviate the need for runtime typing.

Reply:
I accidentally deleted my earlier, but it was moot anyways. At this point, I'm not sure why something like this is necessary in the first place. AFAIK `cnp.intp_t` and `np.inp` should be corresponding types no matter what platform, so there shouldn't be any additional runtime typing we need to take care of. Am I missing something?

Reply:
You are right, not missing something and not required to change anything; I just wondered whether this runtime definition was preferable; it probably isn't.

Reply:
Okay, let's leave it as-is for now then :)